File: examples/examples-kotlin-java/src/test/java/com/mercadolibre/examples/java/UC01.java
package com.mercadolibre.examples.java;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.stream.Collectors;

class State {

	private final String name;


	public State(String name) {
		this.name = name;
	}

	public String getName() {
		return name;
	}
}

class City {

	private final String name;
	private final Integer population;
	private final State state;

	public City(String name, Integer population, State state) {
		this.name = name;
		this.population = population;
		this.state = state;
	}

	public String getName() {
		return name;
	}

	public Integer getPopulation() {
		return population;
	}

	public State getState() {
		return state;
	}
}

class Address {

	private final String street;
	private final City city;

	public Address(String street, City city) {
		this.street = street;
		this.city = city;
	}

	public String getStreet() {
		return street;
	}

	public City getCity() {
		return city;
	}
}

class Person {

	private final String name;
	private final Integer age;
	private final Address address;

	public Person(String name, Integer age, Address address) {
		this.name = name;
		this.age = age;
		this.address = address;
	}

	public String getName() {
		return name;
	}

	public Address getAddress() {
		return address;
	}

	public Integer getAge() {
		return age;
	}
}

class PersonService {

	public static final Integer AGE_LIMIT_FILTER = 18;
	public static final Integer POPULATION_LIMIT_FILTER = 100000;

	public Collection<String> getStates(Collection<Person> persons) {
		if (persons == null) {//VER PARA FAZER COM OPTIONAL
			return new ArrayList<>();
		}
		return persons.stream()
			.filter(person -> person.getAge() != null && person.getAge() >= AGE_LIMIT_FILTER)
			.map(Person::getAddress)
			.map(Address::getCity)
			.filter(city -> city.getPopulation() != null && city.getPopulation() >= POPULATION_LIMIT_FILTER)
			.map(City::getState)
			.map(State::getName)
			.distinct()
			.sorted()
			.collect(Collectors.toList());
	}
}


public class UC01 {

	private PersonService personService = new PersonService();

	@Test
	public void UC01_01() {
		Collection<Person> persons = Arrays.asList(
			new Person("André Justi", null, null)
		);
		Collection<String> states = personService.getStates(persons);
		Assertions.assertTrue(states.isEmpty());
	}

	@Test
	public void UC01_02() {
		Collection<Person> persons = Arrays.asList(
			new Person("André Justi", null, null),
			new Person("Maycow Antunes", null, null)
		);
		Collection<String> states = personService.getStates(persons);
		Assertions.assertTrue(states.isEmpty());
	}

	@Test
	public void UC01_03() {
		Collection<Person> persons = Arrays.asList(
			new Person("Paulo Gustavo", PersonService.AGE_LIMIT_FILTER - 1, new Address("Doutor Carlos Maia", new City("Cosmopolis", PersonService.POPULATION_LIMIT_FILTER + 1, new State("SC"))))
		);
		Collection<String> states = personService.getStates(persons);
		Assertions.assertTrue(states.isEmpty());
	}

	@Test
	public void UC01_04() {
		Collection<Person> persons = Arrays.asList(
			new Person("Wellington Macedo", PersonService.AGE_LIMIT_FILTER + 1, new Address("Miguel Dutra", new City("Palhoça", PersonService.POPULATION_LIMIT_FILTER - 1, new State("SC"))))
		);
		Collection<String> states = personService.getStates(persons);
		Assertions.assertTrue(states.isEmpty());
	}


	@Test
	public void UC01_05() {
		Collection<Person> persons = Arrays.asList(
			new Person("André Justi", PersonService.AGE_LIMIT_FILTER + 10, new Address("Padre Chagas", new City("Florianópolis", PersonService.POPULATION_LIMIT_FILTER + 1, new State("SC"))))
		);
		Collection<String> states = personService.getStates(persons);
		Assertions.assertEquals(Arrays.asList("SC"), states);
	}

	@Test
	public void UC01_06() {
		Collection<Person> persons = Arrays.asList(
			new Person("André Justi", PersonService.AGE_LIMIT_FILTER + 10, new Address("Padre Chagas", new City("Florianópolis", PersonService.POPULATION_LIMIT_FILTER + 1, new State("SC")))),
			new Person("Wellington Macedo", PersonService.AGE_LIMIT_FILTER + 1, new Address("Miguel Dutra", new City("Palhoça", PersonService.POPULATION_LIMIT_FILTER - 1, new State("SC")))),
			new Person("Paulo Gustavo", PersonService.AGE_LIMIT_FILTER + 1, new Address("Jose Angelo Peti", new City("Druta", PersonService.POPULATION_LIMIT_FILTER - 1, new State("SP")))),
			new Person("Maycow Antunes", PersonService.AGE_LIMIT_FILTER + 1, new Address("Dona Ema", new City("Rio Negrinho", PersonService.POPULATION_LIMIT_FILTER + 1, new State("RJ"))))
		);
		Collection<String> states = personService.getStates(persons);
		Assertions.assertEquals(Arrays.asList("RJ", "SC"), states);
	}

	@Test
	public void UC01_07() {
		Collection<Person> persons = Arrays.asList();
		Collection<String> states = personService.getStates(persons);
		Assertions.assertTrue(states.isEmpty());
	}
}

File: examples/examples-kotlin-java/src/test/kotlin/com/mercadolibre/examples/kotlin/UC01.kt
package com.mercadolibre.examples.kotlin

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

data class State(val name: String? = null)
data class City(val name: String? = null, val population: Int? = null, val state: State? = null)
data class Address(val street: String? = null, val city: City? = null)
data class Person(val name: String? = null, val age: Int? = null, val address: Address? = null)

class PersonService {

	companion object {
		const val AGE_LIMIT_FILTER = 18
		const val POPULATION_LIMIT_FILTER = 100000
	}

	fun getStates(persons: Collection<Person>) = persons
		.filter { it.age ?: 0 >= AGE_LIMIT_FILTER }
		.mapNotNull(Person::address)
		.mapNotNull(Address::city)
		.filter { it.population ?: 0 >= POPULATION_LIMIT_FILTER }
		.mapNotNull(City::state)
		.mapNotNull(State::name)
		.distinct()
		.sorted()
}


class UC01 {

	private val personService = PersonService()

	@Test
	fun `UC01_01`() {
		val persons = listOf(
			Person(name = "André Justi", age = null, address = null)
		)
		val states = personService.getStates(persons)
		Assertions.assertTrue(states.isEmpty())
	}

	@Test
	fun `UC01_02`() {
		val persons = listOf(
			Person(name = "André Justi", age = null, address = null),
			Person(name = "Maycow Antunes", age = null, address = null)
		)
		val states = personService.getStates(persons)
		Assertions.assertTrue(states.isEmpty())
	}

	@Test
	fun `UC01_03`() {
		val persons = listOf(
			Person(name = "Paulo Gustavo", age = PersonService.AGE_LIMIT_FILTER - 1, address = Address(street = "Doutor Carlos Maia", city = City(name = "Cosmopolis", population = PersonService.POPULATION_LIMIT_FILTER + 1, state = State(name = "SC"))))
		)
		val states = personService.getStates(persons)
		Assertions.assertTrue(states.isEmpty())
	}

	@Test
	fun `UC01_04`() {
		val persons = listOf(
			Person(name = "Wellington Macedo", age = PersonService.AGE_LIMIT_FILTER + 1, address = Address(street = "Miguel Dutra", city = City(name = "Palhoça", population = PersonService.POPULATION_LIMIT_FILTER - 1, state = State(name = "SC"))))
		)
		val states = personService.getStates(persons)
		Assertions.assertTrue(states.isEmpty())
	}


	@Test
	fun `UC01_05`() {
		val persons = listOf(
			Person(name = "André Justi", age = PersonService.AGE_LIMIT_FILTER + 10, address = Address(street = "Padre Chagas", city = City(name = "Florianópolis", population = PersonService.POPULATION_LIMIT_FILTER + 1, state = State(name = "SC"))))
		)
		val states = personService.getStates(persons)
		Assertions.assertEquals(listOf("SC"), states)
	}

	@Test
	fun `UC01_06`() {
		val persons = listOf(
			Person(name = "André Justi", age = PersonService.AGE_LIMIT_FILTER + 10, address = Address(street = "Padre Chagas", city = City(name = "Florianópolis", population = PersonService.POPULATION_LIMIT_FILTER + 1, state = State(name = "SC")))),
			Person(name = "Wellington Macedo", age = PersonService.AGE_LIMIT_FILTER + 1, address = Address(street = "Miguel Dutra", city = City(name = "Palhoça", population = PersonService.POPULATION_LIMIT_FILTER - 1, state = State(name = "SC")))),
			Person(name = "Paulo Gustavo", age = PersonService.AGE_LIMIT_FILTER + 1, address = Address(street = "Jose Angelo Peti", city = City(name = "Druta", population = PersonService.POPULATION_LIMIT_FILTER - 1, state = State(name = "SP")))),
			Person(name = "Maycow Antunes", age = PersonService.AGE_LIMIT_FILTER + 1, address = Address(street = "Dona Ema", city = City(name = "Rio Negrinho", population = PersonService.POPULATION_LIMIT_FILTER + 1, state = State(name = "RJ"))))
		)
		val states = personService.getStates(persons)
		Assertions.assertEquals(listOf("RJ", "SC"), states)
	}

	@Test
	fun `UC01_07`() {
		val persons = emptyList<Person>()
		val states = personService.getStates(persons)
		Assertions.assertTrue(states.isEmpty())
	}
}

